File: ps5-tv-agent/app/src/main/java/com/bookmygame/tvagent/StationApi.kt
package com.bookmygame.tvagent

import android.util.Log
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * The three station routes the site already exposes.
 *
 * Nothing here is new server work: these are the same endpoints the Windows
 * agent has been using in production, called with the same payloads. A PS5
 * station is just another station as far as the site is concerned.
 */
class StationApi(private val config: AgentConfig) {

    private val http = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(20, TimeUnit.SECONDS)
        // Redirects are followed by default, and following one from the apex to
        // www drops the Authorization header - so the origin is pinned to www
        // and a redirect here means something is wrong rather than routine.
        .followRedirects(false)
        .build()

    private val json = "application/json; charset=utf-8".toMediaType()

    class ApiException(message: String, val status: Int) : Exception(message)

    private fun post(path: String, body: JSONObject, bearer: String? = null): JSONObject {
        val request = Request.Builder()
            .url(config.siteOrigin + path)
            .post(body.toString().toRequestBody(json))
            .apply { bearer?.let { header("Authorization", "Bearer $it") } }
            .build()

        http.newCall(request).execute().use { response ->
            val text = response.body?.string().orEmpty()
            val parsed = try { JSONObject(text) } catch (_: Exception) { JSONObject() }

            if (!response.isSuccessful) {
                val message = parsed.optString("error").ifBlank { "Request failed (${response.code})" }
                throw ApiException(message, response.code)
            }
            return parsed
        }
    }

    /**
     * Trades a setup code generated in the owner dashboard for this station's
     * settings. The code is single use, so a failure here has to say why
     * clearly - the person is standing at the TV with a remote.
     */
    fun enroll(code: String): JSONObject =
        post("/api/stations/enroll", JSONObject().put("code", code))

    /**
     * A short-lived token for the QR on the locked screen.
     *
     * Re-issued well before it expires rather than once at startup: a station
     * sits locked for hours, and a stale code sends a paying customer to a
     * page that refuses them.
     */
    fun unlockToken(): String {
        val body = JSONObject()
            .put("cafeId", config.cafeId)
            .put("stationName", config.stationName)
        return post("/api/stations/unlock-token", body, config.heartbeatToken).getString("token")
    }

    /**
     * Says this station is alive, and what it is showing.
     *
     * The dashboard treats a station that has not reported in 90 seconds as
     * offline, and the QR flow refuses to take money for an offline machine -
     * so this failing is not cosmetic, it stops sales.
     */
    fun heartbeat(status: String, sessionId: String?, version: String) {
        val body = JSONObject()
            .put("cafeId", config.cafeId)
            .put("stationName", config.stationName)
            .put("status", status)
            .put("version", version)
        sessionId?.let { body.put("sessionId", it) }

        try {
            post("/api/stations/heartbeat", body, config.heartbeatToken)
        } catch (err: Exception) {
            // Logged, never thrown: a missed heartbeat must not take the lock
            // down with it. The next one is thirty seconds away.
            Log.w("StationApi", "Heartbeat failed: ${err.message}")
        }
    }
}

File: ps5-pi-agent/agent/station_api.py
"""
The three station routes the site already exposes.

No new server work: these are the same endpoints the Windows agent has used in
production since August, called with the same payloads.
"""

from __future__ import annotations

import logging

import requests

log = logging.getLogger(__name__)


class StationApi:
    def __init__(self, origin: str, station=None) -> None:
        self._origin = origin.rstrip("/")
        self._station = station

    def _post(self, path: str, body: dict, token: str | None = None) -> dict:
        headers = {"Content-Type": "application/json"}
        if token:
            headers["Authorization"] = f"Bearer {token}"

        response = requests.post(
            self._origin + path,
            json=body,
            headers=headers,
            timeout=20,
            # A redirect from the apex to www would silently drop the
            # Authorization header, so a redirect here is a misconfiguration
            # worth seeing rather than following.
            allow_redirects=False,
        )
        parsed = {}
        try:
            parsed = response.json()
        except Exception:
            pass

        if not response.ok:
            raise RuntimeError(parsed.get("error") or f"Request failed ({response.status_code})")
        return parsed

    def enroll(self, code: str) -> dict:
        return self._post("/api/stations/enroll", {"code": code})

    def unlock_token(self) -> str:
        body = {"cafeId": self._station.cafe_id, "stationName": self._station.station_name}
        return self._post("/api/stations/unlock-token", body, self._station.heartbeat_token)["token"]

    def heartbeat(self, status: str, session_id: str | None, version: str) -> None:
        body = {
            "cafeId": self._station.cafe_id,
            "stationName": self._station.station_name,
            "status": status,
            "version": version,
        }
        if session_id:
            body["sessionId"] = session_id
        try:
            self._post("/api/stations/heartbeat", body, self._station.heartbeat_token)
        except Exception as err:
            # Never raised: a missed heartbeat must not take the lock down with
            # it, and the next one is thirty seconds away. It is not cosmetic
            # though - the dashboard calls a station offline after ninety
            # seconds, and the QR flow refuses to sell time on an offline one.
            log.warning("Heartbeat failed: %s", err)
